docs/internals/optimizer.rst

```diff
@@ -206,8 +207,43 @@ The following transformation steps are the main components:
  - Redundant Assign Eliminator
  - Full Function Inliner
 
-Selecting optimizations
-~~~~~~~~~~~~~~~~~~~~~~~
+ Full List of Yul-Based Optimizer Steps
+ -----------------------------------------
+ 
+ - :ref:`BlockFlattener<Block Flattener>`.
+ - CircularReferencesPruner - To be documented.
+ - :ref:`CommonSubexpressionEliminator<Common Subexpression Eliminator>`.
+ - ConditionalSimplifier - To be documented.
+ - ConditionalUnsimplifier - To be documented.
+ - ControlFlowSimplifier - To be documented.
+ - DeadCodeEliminator - To be documented.
+ - :ref:`EquivalentFunctionCombiner<Equivalent Function Combiner>`.
+ - ExpressionInliner - To be documented.
+ - :ref:`ExpressionJoiner<Expression Joiner>`.
+ - :ref:`ExpressionSimplifier<Expression Simplifier>`.
+ - :ref:`ExpressionSplitter<Expression Splitter>`.
+ - :ref:`ForLoopConditionIntoBody<For Loop Condition Into Body>`.
+ - ForLoopConditionOutOfBody - To be documented.
+ - :ref:`ForLoopInitRewriter<For Loop Init Rewriter>`.
+ - :ref:`FullInliner<Full Function Inliner>`.
+ - :ref:`FunctionGrouper<Fuction Grouper>`.
+ - :ref:`FunctionHoister<Function Hoister>`.
+ - FunctionSpecializer - To be documented.
+ - LiteralRematerialiser - To be documented.
+ - LoadResolver - To be documented.
+ - LoopInvariantCodeMotion - To be documented.
+ - :ref:`RedundantAssignEliminator<Redundant Assign Eliminator>`.
+ - ReasoningBasedSimplifier - To be documented.
+ - :ref:`Rematerialiser<Rematerialiser>`.
+ - :ref:`SSAReverser<SSA Reverser>`.
+ - :ref:`SSATransform<SSA Transform>`.
+ - :ref:`StructuralSimplifier<Structural Simplifier>`.
+ - UnusedFunctionParameterPruner - To be documented.
```

Comment:
If you want to add documentation for these, you can copy them from the corresponding header and `.cpp` files. Here `libyul/optimiser/UnusedFunctionParameterPruner.h` and `libyul/optimiser/UnusedFunctionParameterPruner.cpp`. Similarly, for the others in the same folder.

Reply:
So should I do this for all the remaining "to be documented" steps? Also, my question would be where they fit in since we structured them in their sequence as far as I understood? (And some of them seem to not be included in the current Yul optimizer sequence).

Reply:
I'm not sure if it's structured as the sequence. I think, it's okay to have them alphabetically listed.

> So should I do this for all the remaining "to be documented" steps?

Yes, that would be nice. The `optimiser/Readme.md` might be outdated, however the `.h` header and `.cpp` files should contain some explanation about the step. Copying it to the documentation sounds good.

Reply:
They are currently categorized according to a few categories e.g.
- Preprocessing
- Pseudo-SSA Transformation
- Tools
- Statement-Scale Simplifications
...
Can we sort the remaining steps into these categories or should I get rid of those subtitles and put everything in alphabetical order? (For now I added the remaining steps at the end of the doc.)